## Final Project Submission

Please fill out:
* Student name: 
* Student pace: self paced / part time / full time
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


In [ ]:
# Your code here - remember to use markdown cells for comments as well!

In [1]:
ls

CONTRIBUTING.md    README.md          halfway-there.gif
LICENSE.md         data/              student.ipynb


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [10]:
pd.set_option('display.max_columns', 100)

In [11]:
df = pd.read_csv('data/kc_house_data.csv')

In [12]:
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0      5650     1.0         NaN   0.0          3      7        1180   
1      7242     2.0         0.0   0.0          3      7        2170   
2     10000     1.0         0.0   0.0          3      6         770   
3      5000     1.0         0.0   0.0          5      7        1050   
4      8080     1.0         0.0   0.0          3      8        1680   

  sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0           0.0      1955           0.0    98178  47.5112 -122.257   
1         400.0      1951        1991.0    98125  47.7210 -122.319   
2           0.0      1933           NaN    98028  47.7379 -122.233   
3         910.0      1965           0.0    98136  47.5208 -122.393   
4           0.0      1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503

# Column Names and descriptions for Kings County Data Set
* **id** - unique identified for a house
* **dateDate** - house was sold
* **pricePrice** -  is prediction target
* **bedroomsNumber** -  of Bedrooms/House
* **bathroomsNumber** -  of bathrooms/bedrooms
* **sqft_livingsquare** -  footage of the home
* **sqft_lotsquare** -  footage of the lot
* **floorsTotal** -  floors (levels) in house
* **waterfront** - House which has a view to a waterfront
* **view** - Has been viewed
* **condition** - How good the condition is ( Overall )
* **grade** - overall grade given to the housing unit, based on King County grading system
* **sqft_above** - square footage of house apart from basement
* **sqft_basement** - square footage of the basement
* **yr_built** - Built Year
* **yr_renovated** - Year when house was renovated
* **zipcode** - zip
* **lat** - Latitude coordinate
* **long** - Longitude coordinate
* **sqft_living15** - The square footage of interior housing living space for the nearest 15 neighbors
* **sqft_lot15** - The square footage of the land lots of the nearest 15 neighbors


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  float64
 9   view           21534 non-null  float64
 10  condition      21597 non-null  int64  
 11  grade          21597 non-null  int64  
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

**Initial Observation**

sqft_basement is type object - convert to type int

In [27]:
df['sqft_basement'].value_counts()

0.0       12826
?           454
600.0       217
500.0       209
700.0       208
          ...  
1880.0        1
1284.0        1
506.0         1
1960.0        1
602.0         1
Name: sqft_basement, Length: 304, dtype: int64

In [31]:
df['sqft_basement'].value_counts() / len(df['sqft_basement'])

0.0       0.593879
?         0.021021
600.0     0.010048
500.0     0.009677
700.0     0.009631
            ...   
1880.0    0.000046
1284.0    0.000046
506.0     0.000046
1960.0    0.000046
602.0     0.000046
Name: sqft_basement, Length: 304, dtype: float64

There seems to be some '?' as a value for sqft_basement (around 2% of values)

Instead of dropping these, remember that the sqft_basement and sqft_above variables should add up to sqft_living.
So, make sqft_basement for these '?' values to be (sqft_living - sqft_above)

In [46]:
df_drop_questionmark = df.drop(df.loc[df['sqft_basement'] == '?'].index)

In [47]:
df_drop_questionmark['sqft_basement'].value_counts()

0.0       12826
600.0       217
500.0       209
700.0       208
800.0       201
          ...  
2130.0        1
2180.0        1
143.0         1
1284.0        1
1852.0        1
Name: sqft_basement, Length: 303, dtype: int64

In [48]:
df_drop_questionmark['sqft_basement'] = df_drop_questionmark['sqft_basement'].astype('float64').astype('int64')

In [49]:
df_drop_questionmark.corr()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id             1.000000 -0.015796  0.002487   0.006120    -0.011721 -0.132268   
price         -0.015796  1.000000  0.309204   0.525889     0.702328  0.087937   
bedrooms       0.002487  0.309204  1.000000   0.513840     0.577998  0.032262   
bathrooms      0.006120  0.525889  0.513840   1.000000     0.755278  0.087884   
sqft_living   -0.011721  0.702328  0.577998   0.755278     1.000000  0.172941   
sqft_lot      -0.132268  0.087937  0.032262   0.087884     0.172941  1.000000   
floors         0.019281  0.256355  0.179044   0.504071     0.354688 -0.007522   
waterfront    -0.003409  0.278085 -0.000108   0.068293     0.111726  0.022579   
view           0.012859  0.397176  0.080379   0.187249     0.283797  0.072668   
condition     -0.023186  0.035290  0.025710  -0.128015    -0.061294 -0.009671   
grade          0.008681  0.667738  0.356882   0.666493     0.763101  0.113402   
sqft_above    -0.010567  0.605143  0.479210   0.686456     0.876678  0.183461   
sqft_basement -0.004548  0.325008  0.302683   0.282693     0.434576  0.015533   
yr_built       0.022218  0.054459  0.157017   0.508408     0.319461  0.052768   
yr_renovated  -0.010601  0.128227  0.019060   0.051736     0.054291  0.003491   
zipcode       -0.011183 -0.053166 -0.152628  -0.204306    -0.198906 -0.129355   
lat           -0.004728  0.306507 -0.009521   0.024994     0.052986 -0.085457   
long           0.021373  0.022101  0.131093   0.224479     0.240797  0.230583   
sqft_living15 -0.003039  0.586415  0.392476   0.570129     0.756389  0.143428   
sqft_lot15    -0.138005  0.083192  0.030479   0.088834     0.184466  0.720649   

                 floors  waterfront      view  condition     grade  \
id             0.019281   -0.003409  0.012859  -0.023186  0.008681   
price          0.256355    0.278085  0.397176   0.035290  0.667738   
bedrooms       0.179044   -0.000108  0.080379   0.025710  0.356882   
bathrooms      0.504071    0.068293  0.187249  -0.128015  0.666493   
sqft_living    0.354688    0.111726  0.283797  -0.061294  0.763101   
sqft_lot      -0.007522    0.022579  0.072668  -0.009671  0.113402   
floors         1.000000    0.021371  0.027748  -0.264280  0.459214   
waterfront     0.021371    1.000000  0.404847   0.017741  0.088206   
view           0.027748    0.404847  1.000000   0.045464  0.250014   
condition     -0.264280    0.017741  0.045464   1.000000 -0.148508   
grade          0.459214    0.088206  0.250014  -0.148508  1.000000   
sqft_above     0.524225    0.075920  0.166907  -0.159289  0.756382   
sqft_basement -0.245144    0.090200  0.276521   0.170972  0.168023   
yr_built       0.489665   -0.025430 -0.054415  -0.362107  0.448120   
yr_renovated   0.003256    0.083270  0.098948  -0.062261  0.016907   
zipcode       -0.058976    0.030282  0.085300   0.003437 -0.185131   
lat            0.048977   -0.012124  0.006174  -0.015592  0.113380   
long           0.125922   -0.038541 -0.076306  -0.106285  0.199187   
sqft_living15  0.280294    0.090325  0.280934  -0.093493  0.713574   
sqft_lot15    -0.011545    0.031604  0.072797  -0.004860  0.121921   

               sqft_above  sqft_basement  yr_built  yr_renovated   zipcode  \
id              -0.010567      -0.004548  0.022218     -0.010601 -0.011183   
price            0.605143       0.325008  0.054459      0.128227 -0.053166   
bedrooms         0.479210       0.302683  0.157017      0.019060 -0.152628   
bathrooms        0.686456       0.282693  0.508408      0.051736 -0.204306   
sqft_living      0.876678       0.434576  0.319461      0.054291 -0.198906   
sqft_lot         0.183461       0.015533  0.052768      0.003491 -0.129355   
floors           0.524225      -0.245144  0.489665      0.003256 -0.058976   
waterfront       0.075920       0.090200 -0.025430      0.083270  0.030282   
view             0.166907       0.276521 -0.054415      0.098948  0.085300   
condition       -0.159289       0.170972 -0.362107     -0.062261  0.

There seem to be missing values in waterfront and yr_renovated - makes sense because not all houses will have waterfronts or be renovated

In [17]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.159700e+04  21597.000000  19221.000000  21534.000000  21597.000000   
mean   1.509941e+04      1.494096      0.007596      0.233863      3.409825   
std    4.141264e+04      0.539683      0.086825      0.765686      0.650546   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068500e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above      yr_built  yr_renovated       zipcode  \
count  21597.000000  21597.000000  21597.000000  17755.000000  21597.000000   
mean       7.657915   1788.596842   1970.999676     83.636778  98077.951845   
std        1.173200    827.759761     29.375234    399.946414     53.513072   
min        3.000000    370.000000   1900.000000      0.000000  98001.000000   
25%        7.000000   1190.000000   1951.000000      0.000000  98033.000000   
50%        7.000000   1560.000000   1975.000000      0.000000  98065.000000   
75%        8.000000   2210.000000   1997.000000      0.000000  98118.000000   
max       13.000000   9410.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000   21597.000000   21597.000000  
mean      47.560093   -122.213982    1986.620318   12758.283512  
std        0.138552      0.140724     685.230472   27274.441950  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471100   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.231000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

In [15]:
df.corr()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id             1.000000 -0.016772  0.001150   0.005162    -0.012241 -0.131911   
price         -0.016772  1.000000  0.308787   0.525906     0.701917  0.089876   
bedrooms       0.001150  0.308787  1.000000   0.514508     0.578212  0.032471   
bathrooms      0.005162  0.525906  0.514508   1.000000     0.755758  0.088373   
sqft_living   -0.012241  0.701917  0.578212   0.755758     1.000000  0.173453   
sqft_lot      -0.131911  0.089876  0.032471   0.088373     0.173453  1.000000   
floors         0.018608  0.256804  0.177944   0.502582     0.353953 -0.004814   
waterfront    -0.004176  0.276295 -0.002386   0.067282     0.110230  0.023143   
view           0.011592  0.395734  0.078523   0.186451     0.282532  0.075298   
condition     -0.023803  0.036056  0.026496  -0.126479    -0.059445 -0.008830   
grade          0.008188  0.667951  0.356563   0.665838     0.762779  0.114731   
sqft_above    -0.010799  0.605368  0.479386   0.686668     0.876448  0.184139   
yr_built       0.021617  0.053953  0.155670   0.507173     0.318152  0.052946   
yr_renovated  -0.012010  0.129599  0.018495   0.051050     0.055660  0.004513   
zipcode       -0.008211 -0.053402 -0.154092  -0.204786    -0.199802 -0.129586   
lat           -0.001798  0.306692 -0.009951   0.024280     0.052155 -0.085514   
long           0.020672  0.022036  0.132054   0.224903     0.241214  0.230227   
sqft_living15 -0.002701  0.585241  0.393406   0.569884     0.756402  0.144763   
sqft_lot15    -0.138557  0.082845  0.030690   0.088303     0.184342  0.718204   

                 floors  waterfront      view  condition     grade  \
id             0.018608   -0.004176  0.011592  -0.023803  0.008188   
price          0.256804    0.276295  0.395734   0.036056  0.667951   
bedrooms       0.177944   -0.002386  0.078523   0.026496  0.356563   
bathrooms      0.502582    0.067282  0.186451  -0.126479  0.665838   
sqft_living    0.353953    0.110230  0.282532  -0.059445  0.762779   
sqft_lot      -0.004814    0.023143  0.075298  -0.008830  0.114731   
floors         1.000000    0.021883  0.028436  -0.264075  0.458794   
waterfront     0.021883    1.000000  0.406654   0.017642  0.087383   
view           0.028436    0.406654  1.000000   0.045735  0.249727   
condition     -0.264075    0.017642  0.045735   1.000000 -0.146896   
grade          0.458794    0.087383  0.249727  -0.146896  1.000000   
sqft_above     0.523989    0.075463  0.166299  -0.158904  0.756073   
yr_built       0.489193   -0.026079 -0.054564  -0.361592  0.447865   
yr_renovated   0.003535    0.087244  0.100964  -0.061858  0.016669   
zipcode       -0.059541    0.031057  0.085277   0.002888 -0.185771   
lat            0.049239   -0.012772  0.006141  -0.015102  0.113575   
long           0.125943   -0.039864 -0.077894  -0.105877  0.200341   
sqft_living15  0.280102    0.088860  0.279561  -0.093072  0.713867   
sqft_lot15    -0.010722    0.032002  0.073332  -0.003126  0.120981   

               sqft_above  yr_built  yr_renovated   zipcode       lat  \
id              -0.010799  0.021617     -0.012010 -0.008211 -0.001798   
price            0.605368  0.053953      0.129599 -0.053402  0.306692   
bedrooms         0.479386  0.155670      0.018495 -0.154092 -0.009951   
bathrooms        0.686668  0.507173      0.051050 -0.204786  0.024280   
sqft_living      0.876448  0.318152      0.055660 -0.199802  0.052155   
sqft_lot         0.184139  0.052946      0.004513 -0.129586 -0.085514   
floors           0.523989  0.489193      0.003535 -0.059541  0.049239   
waterfront       0.075463 -0.026079      0.087244  0.031057 -0.012772   
view             0.166299 -0.054564      0.100964  0.085277  0.006141   
condition       -0.158904 -0.361592     -0.061858  0.002888 -0.015102   
grade            0.756073  0.447865      0.016669 -0.185771  0.113575   
sqft_above       1.000000  0.424037      0.022137 -0.261570 -0.001199   
yr_built         0.424037  1.000000     -0.225209 

sqft_living and sqft_above have a high colinearity of .876 -> drop 